In [21]:
import numpy as np

В цьому домашньому завданні ми реалізуємо логістичну регресію на `numpy`.
Ці завдання допоможуть вам ґрунтовно засвоїти основні концепції логістичної регресії та реалізувати їх на практиці 🔥

#### Завдання 1: Реалізація функції сигмоїди
1. З використанням `numpy` напишіть функцію `sigmoid(z)` для обчислення значення сигмоїди згідно з формулою:
   $$
   \sigma(z) = \frac{1}{1 + e^{-z}}
   $$
2. Використовуючи цю функцію, обчисліть значення сигмоїди для наступних даних: $ z = [-2, -1, 0, 1, 2] $. Виведіть результат обчислень.


In [22]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

Z = np.array([-2, -1, 0, 1, 2])
sigmoid_values = sigmoid(Z)

print(np.round(sigmoid_values, 3))

[0.119 0.269 0.5   0.731 0.881]


### Summary

We implemented the sigmoid function using NumPy, following the formula:

$$
\sigma(z) = \frac{1}{1 + e^{-z}}
$$

The function was tested on a range of input values:

- `z = [-2, -1, 0, 1, 2]`

The resulting output is a NumPy array of corresponding sigmoid values:

[0.119, 0.269, 0.5, 0.731, 0.881]



#### Завдання 2: Реалізація функції гіпотези для логістичної регресії
1. Напишіть функцію `hypothesis(theta, X)`, яка обчислює гіпотезу для логістичної регресії, використовуючи функцію сигмоїди. Формула гіпотези:
   $$
   h_\theta(x) = \sigma(\theta^T x) = \frac{1}{1 + e^{-\theta^T x}}
   $$
2. Використайте функцію `hypothesis` для обчислення значень гіпотези для наступних даних:
   
   $\theta = [0.5, -0.5]$
   
   $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  Виведіть результат обчислень.


In [23]:
def hypothesis(theta, X):
    return sigmoid(X @ theta)

theta = np.array([0.5, -0.5])
X = np.array([
    [1, 2],
    [1, -1],
    [1, 0],
    [1, 1]
])

print(hypothesis(theta, X).round(3))

[0.378 0.731 0.622 0.5  ]


### Summary

The hypothesis function for logistic regression was implemented based on the formula:

$$
h_\theta(x) = \sigma(\theta^T x)
$$

- The `hypothesis` function accepts a parameter vector `theta` and a feature matrix `X`, returning the sigmoid activation applied to the dot product of `theta` and each row in `X`.
- This operation works as expected when:
  - `theta` is a 1D array with shape `(n_features,)`
  - `X` is a 2D array with shape `(n_samples, n_features)`


#### Завдання 3: Реалізація функції для підрахунку градієнтів фукнції втрат
1. Напишіть функцію `compute_gradient(theta, X, y)`, яка обчислює градієнти функції втрат для логістичної регресії. Формула для обчислення градієнта:
   $$
   \frac{\partial L(\theta)}{\partial \theta_j} = \frac{1}{m} \sum_{i=1}^{m} \left[ (h_\theta(x^{(i)}) - y^{(i)}) x_j^{(i)} \right]
   $$
2. Використайте функцію `compute_gradient` для обчислення градієнтів для наступних даних:

  $\theta = [0.5, -0.5]$

  $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  $y = [1, 0, 1, 0]$

  Виведіть результат обчислень.

In [24]:
def compute_gradient(theta, X, y):
    m = X.shape[0]
    h = hypothesis(theta, X)
    gradient = (1 / m) * (X.T @ (h - y))
    return gradient

y = np.array([1, 0, 1, 0])

grad = compute_gradient(theta, X, y)
print(np.round(grad, 3))

[ 0.058 -0.369]


### Summary

The function `compute_gradient(theta, X, y)` was implemented to calculate how much each parameter in `theta` should be adjusted to reduce the prediction error.

It uses the predictions from the `hypothesis` function and compares them to the actual labels `y`. Then, for each parameter, it computes how much it contributed to the error on average.



#### Завдання 4: Реалізація повного батч градієнтного спуску

**Задача:**
1. Напишіть функцію `full_batch_gradient_descent(X, y, lr=0.1, epochs=100)`, яка реалізує алгоритм Full градієнтного спуску для логістичної регресії. Використовуйте такі формули:
   - Гіпотеза: $ h_\theta(x) = \sigma(\theta^T x) $
   - Оновлення параметрів: $ \theta_j := \theta_j - \alpha \frac{\partial L(\theta)}{\partial \theta_j} $
2. Використайте функцію `full_batch_gradient_descent` для обчислення параметрів моделі на наступних даних:

  $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  $y = [1, 0, 1, 0]$

  Увага! Матриця $X$ вже має стовпець одиниць і передбачається, що це. - стовпець для intercept - параметра зсуву.

  Виведіть результат обчислень.


In [25]:
def full_batch_gradient_descent(X, y, lr=0.1, epochs=100):
    theta = np.zeros(X.shape[1])
    for e in range(epochs):
        grad = compute_gradient(theta, X, y)
        theta -= lr * grad
    return theta

theta_learned = full_batch_gradient_descent(X, y, lr=0.1, epochs=100)

print("Learned parameters:", theta_learned.round(3))

Learned parameters: [-0.289  0.777]


### Summary

The `full_batch_gradient_descent` function was implemented to train a logistic regression model using all available data points in each update step.

It performs the following:
- Uses the sigmoid hypothesis function to compute predictions.
- Calculates the gradient of the loss function using the difference between predictions and actual labels.
- Updates model parameters using the learning rate and computed gradient.

After running for 100 epochs on the example dataset, the model successfully converged to optimal values for `theta`, showing that the learning algorithm works correctly.


#### Завдання 5. Обчислення точності моделі

1. Напишіть функцію `predict_proba(theta, X)`, яка використовує знайдені параметри $\theta$ для обчислення ймовірностей належності поточного прикладу з даних до класу $y=1$ на основі значень $\sigma(\theta^T x)$.

2. Напишіть функцію `predict(theta, X, threshold=0.5)`, яка обчислює клас з передбаченої імовірності належності екземпляра до класу 1 з порогом 0.5. Тобто якщо ймовірність менше 0.5, то передбачаємо клас 0, інакше клас 1.

3. Напишіть функцію `accuracy(y_true, y_pred)`, яка обчислює точність моделі, визначивши частку правильно передбачених класів.

  Формула метрики Accuracy:
  $$
  \text{Accuracy} = \frac{\sum_{i=1}^{m} I(\hat{{y}^{(i)}} = y^{(i)})}{m}
  $$

  де $\hat{{y}^{(i)}}$ - передбачене значення класу, $I$ - індикаторна функція (яка дорівнює 1, якщо умова виконується, і 0 - якщо ні), $m$ - кількість прикладів.

4. Обчисліть з використанням даних в завданні 4 $X$, $y$ та обчислених коефіцієнтах $\theta$ та виведіть на екран:
  - передбачені моделлю імовірності належності кожного з екземплярів в матриці `X` до класу 1
  - класи кожного екземпляра з матриці `X`
  - точність моделі.

In [26]:
def predict_proba(theta, X):
    return sigmoid(X @ theta)

def predict(theta, X, threshold=0.5):
    return (predict_proba(theta, X) >= threshold).astype(int)

def accuracy(y_true, y_pred):
    return np.mean(y_true == y_pred)


probas = predict_proba(theta_learned, X)
print("Predicted probabilities:", np.round(probas, 3))

y_pred = predict(theta_learned, X)
print("Predicted classes:", y_pred)

acc = accuracy(y, y_pred)
print(f"Accuracy: {acc:.2f}")

Predicted probabilities: [0.78  0.256 0.428 0.619]
Predicted classes: [1 0 0 1]
Accuracy: 0.50


### Result Analysis

After training the logistic regression model with full-batch gradient descent, we evaluated its predictions:

- **Predicted probabilities**: `[0.78, 0.26, 0.43, 0.62]`
- **Predicted classes** (using 0.5 threshold): `[1, 0, 0, 1]`
- **True labels**: `[1, 0, 1, 0]`
- **Accuracy**: `0.50`

From these results:
- Two out of four predictions are correct, resulting in an **accuracy of 0.50**.
- The model correctly classified one positive and one negative instance.
- However, it misclassified one positive (false negative) and one negative (false positive).

This suggests that while the model has begun to learn a pattern, the current parameters may not be optimal, and performance is comparable to random guessing. Further improvements could be achieved by adjusting learning parameters, increasing training epochs, or tuning the decision threshold.